Add Green Context CUDA kernel launch-start time benchmark

chengronglai opened the pull request:
- New HoloHub benchmark application for measuring CUDA kernel launch-start time improvements with NVIDIA Green Context technology
- Uses CUPTI for precise GPU timing measurements under realistic contention
- Includes controlled A/B testing (baseline vs Green Context partitions)
- Comprehensive documentation with usage, troubleshooting, and tuning guides
- Supports configurable load intensity, workload size, and thread configurations

<!-- greptile_comment -->

<h2>Greptile Overview</h2>

Updated On: 2025-10-20 14:23:36 UTC

### Greptile Summary

This PR introduces a new benchmark application to HoloHub that measures CUDA kernel launch-start time improvements when using NVIDIA Green Context technology. Green Context is a GPU resource partitioning feature that reduces kernel scheduling contention. The benchmark implements controlled A/B testing by creating two Holoscan operators: `DummyLoadOp` (generates realistic GPU contention via background load kernels) and `TimingBenchmarkOp` (measures timing of simple benchmark kernels using CUPTI profiling). The implementation compares baseline performance (non-default streams without Green Context) against partitioned Green Context streams, collecting percentile-based statistics to quantify tail latency improvements. The benchmark integrates with HoloHub's build system via CMakeLists additions, includes comprehensive documentation with platform-specific usage guides and extensive performance matrices, and provides configurable parameters for load intensity, workload size, and thread configurations.

### Important Files Changed

<details>
<summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| benchmarks/CMakeLists.txt | 5/5 | Registers new `green_context_benchmarking` application with build system |
| benchmarks/green_context_benchmarking/metadata.json | 5/5 | Adds standard HoloHub metadata defining benchmark properties and requirements |
| benchmarks/green_context_benchmarking/benchmark_cuda_kernel.cu.hpp | 4.5/5 | Declares CUDA kernels and async launcher functions for benchmarking |
| benchmarks/green_context_benchmarking/cupti_profiler.hpp | N/A | File content not provided in diff - cannot review |
| benchmarks/green_context_benchmarking/README.md | 3/5 | Comprehensive documentation with critical syntax errors in command examples |
| benchmarks/green_context_benchmarking/CMakeLists.txt | 3.5/5 | Build configuration with redundant CUDA enabling and unverified CCCL fallback |
| benchmarks/green_context_benchmarking/benchmark_cuda_kernel.cu | 3/5 | CUDA kernel implementations with uninitialized memory reads |
| benchmarks/green_context_benchmarking/green_context_benchmark.cpp | 3/5 | Main benchmark application with format string bugs in logging statements |

</details>

### Confidence score: 2.5/5

- This PR requires careful review before merging due to several bugs that will cause runtime failures or incorrect output
- Score reflects critical issues including format string mismatches causing crashes (green_context_benchmark.cpp lines 442-443, 458-460), missing closing quotes in README command examples that will fail when copy-pasted (lines 58, 97-98), uninitialized memory reads in CUDA kernels (benchmark_cuda_kernel.cu line 42), and potential redundant/unverified build configuration (CMakeLists.txt CUDA language double-enable and CCCL fallback paths)
- Files requiring immediate attention: green_context_benchmark.cpp (fix format strings), README.md (fix command syntax), benchmark_cuda_kernel.cu (initialize memory or document pre-initialization requirement), CMakeLists.txt (verify CCCL fallback behavior)

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant Main
    participant CuptiProfiler as CuptiSchedulingProfiler
    participant BenchmarkApp as GreenContextBenchmarkApp
    participant TimingOp as TimingBenchmarkOp
    participant DummyLoadOp
    participant CUDA
    participant CUPTI

    User->>Main: Run benchmark with args
    Main->>Main: Parse command line arguments
    Main->>CuptiProfiler: getInstance()
    Main->>CuptiProfiler: initialize()
    CuptiProfiler->>CUPTI: cuptiSubscribe()
    CuptiProfiler->>CUPTI: cuptiEnableDomain()
    CuptiProfiler->>CUPTI: cuptiActivityEnable()
    CuptiProfiler->>CUPTI: cuptiActivityRegisterCallbacks()
    
    alt mode == "baseline" or "all"
        Main->>BenchmarkApp: Create app (use_green_context=false)
        BenchmarkApp->>BenchmarkApp: compose()
        BenchmarkApp->>DummyLoadOp: Create with stream pool
        BenchmarkApp->>TimingOp: Create with stream pool
        Main->>BenchmarkApp: run()
        
        loop total_samples times
            BenchmarkApp->>DummyLoadOp: compute()
            DummyLoadOp->>CUDA: allocate_cuda_stream()
            DummyLoadOp->>CUDA: async_run_background_load_kernel()
            DummyLoadOp->>CUDA: cudaStreamSynchronize()
            
            par Parallel execution
                BenchmarkApp->>TimingOp: compute()
                TimingOp->>CUDA: allocate_cuda_stream()
                TimingOp->>CUDA: async_run_simple_benchmark_kernel()
                CUPTI->>CuptiProfiler: apiCallback() (launch timestamp)
                CUDA->>CUDA: Execute kernel
                CUPTI->>CuptiProfiler: bufferCompleted() (activity records)
                TimingOp->>CUDA: cudaStreamSynchronize()
                TimingOp->>CuptiProfiler: cuptiActivityFlushAll()
                TimingOp->>CuptiProfiler: getLatestSchedulingLatency()
                TimingOp->>CuptiProfiler: getLatestExecutionDuration()
            end
        end
        
        Main->>BenchmarkApp: get_cuda_kernel_launch_start_time_benchmark_stats()
        Main->>BenchmarkApp: get_cuda_kernel_execution_time_benchmark_stats()
        Main->>BenchmarkApp: get_dummy_load_execution_time_benchmark_stats()
    end
    
    alt mode == "green-context" or "all"
        Main->>BenchmarkApp: Create app (use_green_context=true)
        BenchmarkApp->>BenchmarkApp: compose()
        BenchmarkApp->>CUDA: cudaGetDeviceProperties()
        BenchmarkApp->>BenchmarkApp: Create CudaGreenContextPool with partitions
        BenchmarkApp->>DummyLoadOp: Create with green context partition 0
        BenchmarkApp->>TimingOp: Create with green context partition 1
        Main->>BenchmarkApp: run()
        
        loop total_samples times
            BenchmarkApp->>DummyLoadOp: compute()
            DummyLoadOp->>CUDA: allocate_cuda_stream() (partition 0)
            DummyLoadOp->>CUDA: async_run_background_load_kernel()
            DummyLoadOp->>CUDA: cudaStreamSynchronize()
            
            par Parallel execution (isolated partitions)
                BenchmarkApp->>TimingOp: compute()
                TimingOp->>CUDA: allocate_cuda_stream() (partition 1)
                TimingOp->>CUDA: async_run_simple_benchmark_kernel()
                CUPTI->>CuptiProfiler: apiCallback() (launch timestamp)
                CUDA->>CUDA: Execute kernel (isolated)
                CUPTI->>CuptiProfiler: bufferCompleted() (activity records)
                TimingOp->>CUDA: cudaStreamSynchronize()
                TimingOp->>CuptiProfiler: cuptiActivityFlushAll()
                TimingOp->>CuptiProfiler: getLatestSchedulingLatency()
                TimingOp->>CuptiProfiler: getLatestExecutionDuration()
            end
        end
        
        Main->>BenchmarkApp: get_cuda_kernel_launch_start_time_benchmark_stats()
        Main->>BenchmarkApp: get_cuda_kernel_execution_time_benchmark_stats()
        Main->>BenchmarkApp: get_dummy_load_execution_time_benchmark_stats()
    end
    
    Main->>Main: calculate_benchmark_stats()
    Main->>Main: print_comprehensive_timing_results()
    Main->>Main: Print comparison (if mode == "all")
    Main->>CuptiProfiler: cleanup()
    CuptiProfiler->>CUPTI: cuptiActivityDisable()
    CuptiProfiler->>CUPTI: cuptiUnsubscribe()
    Main->>User: Display benchmark results
```

<!-- greptile_other_comments_section -->

<!-- /greptile_comment -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a CUDA benchmarking application to measure kernel launch-to-start latency and execution time with baseline vs. green-context modes; configurable samples, workload size, threads-per-block, and load intensity.
  * Includes CUDA kernels, async launch helpers, and CUPTI-based scheduling and execution-time profiling to evaluate contention and partitioning effects.

* **Documentation**
  * Added comprehensive README and metadata with usage instructions, command-line options, sample output, platform notes, and result interpretation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chengronglai committed:
Add Green Context CUDA kernel launch-start time benchmark

- New HoloHub benchmark application for measuring CUDA kernel launch-start
  time improvements with NVIDIA Green Context technology
- Uses CUPTI for precise GPU timing measurements under realistic contention
- Includes controlled A/B testing (baseline vs Green Context partitions)
- Comprehensive documentation with usage, troubleshooting, and tuning guides
- Supports configurable load intensity, workload size, and thread configurations

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai committed:
Fix lint issues

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

sohamm17 reviewed (changes requested):
Minor changes.

chengronglai committed:
Enhance green context benchmarking with safety fixes and comprehensive results

* Fix formatting in cupti_profiler.hpp
* Remove unused context param from run_background_load_kernel
* Add division by zero protection in background load kernel
* Add executive summary with RTX 6000 Ada Generation and Orin results

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/green_context_benchmarking/cupti_profiler.hpp:
```diff
@@ -0,0 +1,285 @@
+        instance_ = new CuptiSchedulingProfiler();
+    }
+    return instance_;
+    }
```
done

chengronglai commented on benchmarks/green_context_benchmarking/benchmark_cuda_kernel.cu:
```diff
@@ -0,0 +1,71 @@
+
+    // Add memory access patterns
+    if (i % 10 == 0) {
+      value += data[(idx + i) % size] * 0.001f;
```
done

chengronglai committed:
Add kernel execution time measurement to green context benchmark

* Extended CUPTI profiler to track execution duration alongside launch - start
* Added execution_durations_ map to store GPU-side kernel execution times
* Added kernel execution time comparison in performance analysis section and results
  display
* Update Sample Output to include Kernel Execution Time

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

tbirdso commented:
Discussed with @sohamm17 , review is in progress and comments will be available in a couple of weeks.

chengronglai committed:
Add Jetson Thor support and CUDA 13 compatibility for green context benchmarking

* Add CCCL configuration in CMakeLists.txt for CUDA 13 header compatibility
* Update README with platform-specific usage instructions for x86_64/Orin vs modern GPUs/Thor
* Add troubleshooting section if encounter compatibility issue with CUDA 13
* Add Jetson Thor performance matrix

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

sohamm17 reviewed (changes requested):
minor changes. otherwise, LGTM.

chengronglai committed:
Improve Green Context benchmark documentation with disclaimers and troubleshooting

- Add CUPTI overhead disclaimers and validation findings
- Clarify measurements are NOT official NVIDIA CUDA launch latency specifications
- Update GPU compatibility guidance: rename to IGX/Jetson, add CUDA < 13 guidance
- Improve troubleshooting for build cache contamination when switching container images

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

sohamm17 reviewed (approved):
LGTM

bhashemian committed:
Merge branch 'main' into benchmarks/green_context_benchmark_refactor

chengronglai commented on benchmarks/green_context_benchmarking/cupti_profiler.hpp:
CPU launch should always precede GPU execution

chengronglai commented on benchmarks/green_context_benchmarking/benchmark_cuda_kernel.cu:
```diff
@@ -0,0 +1,71 @@
+  if (idx >= size || size == 0)
+    return;
+
+  float value = data[idx];
```
DummyLoopOp properly initializes the data

chengronglai commented on benchmarks/green_context_benchmarking/cupti_profiler.hpp:
```diff
@@ -0,0 +1,315 @@
+  // CUPTI activity buffer management - increased buffer size for high contention
+  static void CUPTIAPI bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords) {
+    *size = 64 * 1024;  // Increased to 64KB buffer for better handling under load
+    *buffer = (uint8_t*)malloc(*size);
```
When malloc fails: malloc returns nullptr which is handled later

chengronglai committed:
Correct format string mismatches and add const qualifiers to kernel parameters

* Fix format string in GPU info logging
* Fix format string in green context configuration logging
* Add const qualifiers to size and intensity parameters in CUDA kernel
* Add logging to avoid silently discard latency measurements
* Remove variables never used

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/green_context_benchmarking/benchmark_cuda_kernel.cu.hpp:
```diff
@@ -0,0 +1,30 @@
+// CUDA kernel declarations for benchmarking
+__global__ void simple_benchmark_kernel(float* data, int size);
+
+__global__ void background_load_kernel(float* data, int size, int intensity);
```
done

chengronglai commented on benchmarks/green_context_benchmarking/cupti_profiler.hpp:
```diff
@@ -0,0 +1,315 @@
+
+    try {
+      // Subscribe to CUPTI callbacks
+      CUptiResult result = cuptiSubscribe(&subscriber_, (CUpti_CallbackFunc)apiCallback, this);
```
this is legitimate use of C-style casts

bhashemian committed:
Merge branch 'main' into benchmarks/green_context_benchmark_refactor

bhashemian committed:
Merge branch 'main' into benchmarks/green_context_benchmark_refactor

bhashemian reviewed (approved):
LGTM!

chengronglai committed:
Add error checking and fix portability fix

* Guard global declarations and include CUDA stream type.
* Add error checking for CUDA API call.
* Remove enable_language(CUDA) in CMake as it's automitically enabled
* Add missing standard library in cupti_profiler.hpp
* Broader lock scope to prevent race condition

Signed-off-by: Cheng-Rong Lai <chengrongl@nvidia.com>

chengronglai commented on benchmarks/green_context_benchmarking/green_context_benchmark.cpp:
done

chengronglai commented on benchmarks/green_context_benchmarking/cupti_profiler.hpp:
```diff
@@ -0,0 +1,311 @@
+
+#include <cupti.h>
+#include <unordered_map>
+#include <mutex>
```
done

bhashemian committed:
Merge branch 'main' into benchmarks/green_context_benchmark_refactor
